### ISSUE TITLE
Popup size and orientation with few candidates

### ISSUE THREAD

**Issue by gusbrs** [minad/corfu#17]
Hi @minad , back again.  ;-)

I'm playing with the new child-frame implementation, and it is looking much more robust than the overlay. Thanks!

But I did find some corner cases, mostly related to few candidates, which I bring to your attention. (Tests made with Emacs 27.2 and `corfu` commit `b8cd63f`).

You certainly already know, but to start, the popup does not (always?) shrink when there are less candidates than its height:

![Screenshot from 2021-04-29 10-59-19](https://user-images.githubusercontent.com/19410606/116563777-b9e40f00-a8da-11eb-9505-5f4dc9f70b7f.png)

On v0.4 when the popup was displayed upwards that would mean the candidates got away from point. This is no longer the case in the current version. But there are still issues when things change during completion. For example, start the same completion as above, but in the bottom of the window:

![Screenshot from 2021-04-29 11-11-13](https://user-images.githubusercontent.com/19410606/116564769-a8e7cd80-a8db-11eb-9dc8-61ac1683bda3.png)

Looking good.

Now type `DEL` (`backward-delete-char-untabify` in this case):

![Screenshot from 2021-04-29 11-12-11](https://user-images.githubusercontent.com/19410606/116565025-da609900-a8db-11eb-9fcd-dd6ed1cbb669.png)

Another case. Start completion from a centered position, such as the one in the first screenshot above. Now call `recenter-top-bottom` until point goes to `bottom`, the result is:

![Screenshot from 2021-04-29 11-15-39](https://user-images.githubusercontent.com/19410606/116565462-46db9800-a8dc-11eb-9d22-261e9aeef24d.png)

I know well it's all fresh. I'm just bringing the cases that come about, as you requested me to do, so that it can be made more robust.

**Comment by minad** [comment]
The non-resizing popup seems to be an Emacs bug. In my case I sometimes observe some flicker but then the popup redraws correctly. You could randomly scatter `redisplay` or `redraw-frame` calls somewhere. The issue where the popup is drawn at the wrong position could also be an Emacs bug or some interaction with some other setting. No idea.

**Comment by minad** [comment]
Are you using the Emacs with X/Gtk?

**Comment by gusbrs** [comment]
> The non-resizing popup seems to be an Emacs bug. In my case I sometimes observe some flicker but then the popup redraws correctly.

I don't observe flickering or delay, those screenshots are stable situations for me.

> Are you using the Emacs with X/Gtk?

Do you mean the gtk branch? To be specific in the answer, I built v 27.2 from the release tarball, on a Linux Mint 20.1 Cinnamon, using configuration options `./configure --with-mailutils --with-xwidgets --with-modules --with-cairo` which results for me in `system-configuration-features` being `"XPM JPEG TIFF GIF PNG RSVG CAIRO SOUND GPM DBUS GSETTINGS GLIB NOTIFY INOTIFY ACL LIBSELINUX GNUTLS LIBXML2 FREETYPE HARFBUZZ M17N_FLT LIBOTF ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS XWIDGETS LIBSYSTEMD JSON PDUMPER LCMS2 GMP"`.

> I cannot reproduce any of those issues.

I did not use `emacs -Q`, would you prefer I test it under stricter conditions?

**Comment by minad** [comment]
Yes, can you reproduce this with emacs -Q?

**Comment by gusbrs** [comment]
Mostly yes. Namely the first two cases.

Starting with `emacs -Q`: 

```emacs-lisp
(add-to-list 'load-path "~/.emacs.d/lib/corfu")
(load "corfu.el" nil t t)
(corfu-mode)
```
As far as I get, as vanilla as it gets. I didn't even let `package.el` do its thing.

Screenshots for the first two cases:

![Screenshot from 2021-04-29 12-51-00](https://user-images.githubusercontent.com/19410606/116580743-0c78f780-a8ea-11eb-9350-f166f61cbe4c.png)

![Screenshot from 2021-04-29 12-51-38](https://user-images.githubusercontent.com/19410606/116580763-113dab80-a8ea-11eb-966a-ab32a3769807.png)

The `recenter-top-bottom` case looks good though:

![Screenshot from 2021-04-29 12-52-34](https://user-images.githubusercontent.com/19410606/116580829-20bcf480-a8ea-11eb-953e-b400dde33d5e.png)

**Comment by minad** [comment]
Interesting. I don't see this on my system, neither with my config nor in emacs -Q. This seems to be an Emacs bug. I cannot do much about this. Either the character size reported by Emacs is not correct (leading to a wrong size computation) or the Emacs display simply ignores my frame resize/repositioning calls.

**Comment by gusbrs** [comment]
A guess: are you running on 28?

Either way, I get it. But I was bound to report it.  ;-)

**Comment by minad** [comment]
No, I am also using 27 mostly. I just tested 28 and found another issue with the border. But besides that it works for me on both 27 and 28 as is.

**Comment by gusbrs** [comment]
I really don't know what might be at play, it's out of my depth. Sorry.

**Comment by minad** [comment]
> I really don't know what might be at play, it's out of my depth. Sorry.

Well, it is a bug in the display code of Emacs. That's at play here. But I will have a hard time to report an Emacs bug if I cannot even reproduce it here-. Hopefully it will go away with the pgtk backend :shrug: 

Maybe @tumashu who wrote posframe has seen something like this?

**Comment by gusbrs** [comment]
> Well, it is a bug in the display code of Emacs. That's at play here. But I will have a hard time to report an Emacs bug if I cannot even reproduce it here-. Hopefully it will go away with the pgtk backend.

I understand, hopefully it doesn't bite many people. Thank you for looking into it.

And if you do have any specific suspects on things which might be at play, I can of course provide further system information, as you need it.

**Comment by minad** [comment]
@gusbrs Can you please give it another try with 03e2318. There seems to be a resizing bug on gtk3/gnome which can be worked around by setting some magic variable.

**Comment by gusbrs** [comment]
Hi @minad , thank you for bearing with the issue.  But, no, it doesn't quite fix it.  I can still generate the following:

![Screenshot from 2021-04-30 11-22-34](https://user-images.githubusercontent.com/19410606/116709395-01d06800-a9a7-11eb-9304-db904a97dba9.png)
![Screenshot from 2021-04-30 11-24-54](https://user-images.githubusercontent.com/19410606/116709413-04cb5880-a9a7-11eb-8a9a-3792c5255673.png)
![Screenshot from 2021-04-30 11-25-26](https://user-images.githubusercontent.com/19410606/116709421-06951c00-a9a7-11eb-87fc-7b6fc4b0a343.png)
![Screenshot from 2021-04-30 11-26-07](https://user-images.githubusercontent.com/19410606/116709428-085edf80-a9a7-11eb-8e76-00c1d69942c2.png)

**Comment by gusbrs** [comment]
With

```
(gtk-resize-child-frames 'resize-mode)
```
I get some improvement, but the issues remain. So yes, the environment variables tests must be somehow failing (which see).

With

```
(gtk-resize-child-frames 'hide)
```
I could discern no observable change from what happened previously.

Output of `system-configuration-features`: `"XPM JPEG TIFF GIF PNG RSVG CAIRO SOUND GPM DBUS GSETTINGS GLIB NOTIFY INOTIFY ACL LIBSELINUX GNUTLS LIBXML2 FREETYPE HARFBUZZ M17N_FLT LIBOTF ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS XWIDGETS LIBSYSTEMD JSON PDUMPER LCMS2 GMP"`.

Output of `(getenv "XDG_CURRENT_DESKTOP")`: `"X-Cinnamon"`.

Output of `(getenv "DESKTOP_SESSION")`: `"cinnamon"`.

So, indeed, the checks for `"GNOME"` fail here, but still, though `(gtk-resize-child-frames 'resize-mode)` improves things, it doesn't quite solve them.

**Comment by gusbrs** [comment]
> What does that mean?

I was about to write again. It means I didn't check well, I can still generate all four screenshots above with `(gtk-resize-child-frames 'resize-mode)`...

**Comment by minad** [comment]
Okay, I see. I give up. Please use a proper window manager. What about i3?

**Comment by minad** [comment]
It is unbelievable how bad the current display infrastructure is. There is a reason why I originally started with overlays :(

**Comment by gusbrs** [comment]
> Okay, I see. I give up.

Thanks again.

> Please use a proper window manager. What about i3?

When I grow up, I'll try!  :smile:

**Comment by gusbrs** [comment]
> It is unbelievable how bad the current display infrastructure is. There is a reason why I originally started with overlays :(

Frustrating indeed...

**Comment by gusbrs** [comment]
I copy-pasted from your comment, let me try again.

**Comment by tumashu** [comment]
another way is using pos-tip, which use tooltip, tooltip has a variable, which cache tooltip and let tooltip show hide faster

**Comment by gusbrs** [comment]
Checked with the new commit, and it's looking good here how!  Thanks a million!

For your pleasure, and for the record, the same cases now (one of them doesn't even make sense any longer, since it depended on a previous wrong resizing):

![Screenshot from 2021-04-30 12-09-09](https://user-images.githubusercontent.com/19410606/116715683-4f4fd380-a9ad-11eb-9d26-291485762055.png)
![Screenshot from 2021-04-30 12-09-51](https://user-images.githubusercontent.com/19410606/116715697-51199700-a9ad-11eb-92ce-22eabd411032.png)
![Screenshot from 2021-04-30 12-10-12](https://user-images.githubusercontent.com/19410606/116715704-52e35a80-a9ad-11eb-9463-5dc54890f79a.png)

**Comment by gusbrs** [comment]
One small comment on the commit. It is probably wise to let bind `case-fold-search` to t there, to make those string matches robust to possible different values of the variable.

**Comment by minad** [comment]
Okay, what a relief :sweat_smile: Agree, it is a good idea to set `case-fold-search`.
